flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/InternalReadOnlyKeyContext.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.heap;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.state.KeyGroupRange;
+
+/**
+ * This interface is the current context of a keyed state. It provides information about the
+ * currently selected key in the context, the corresponding key-group, and other key and
+ * key-grouping related information.
+ */
+@Internal
+public interface InternalReadOnlyKeyContext<K> {
```

Comment:
Why we have to introduce `InternalReadOnlyKeyContext`. Is it possible to just use `InternalKeyContext` instead?


We can use the InternalKeyContext maintained in `AbstractKeyedStateBackend` to initiate `StateChangeLogger`, so that we do not need to have AbstractKeyedStateBackend implement InternalReadOnlyKeyContext?

Reply:
Removed `InternalReadOnlyKeyContext`.

I wanted to prevent unnecessary access to changing the current key. But later during recovery implementation I realized I have to pass `InternalKeyContext` anyways.